### ISSUE TITLE
Baseline Results not Matching MemoCMT Paper

### ISSUE THREAD

**Issue by CaydenSchalk** [namphuongtran9196/MemoCMT#2]
I’m trying to reproduce the baseline MemoCMT results from the paper using this repository, but I am getting substantially lower performance than what is reported.

I think one possible source of the discrepancy is that the paper reports 5-fold speaker-independent cross-validation, whereas the released repository appears to use a single fixed train/validation/test split instead. However, the reproduced performance is still substantially lower than the reported results, so I suspect there may also be differences in preprocessing, split construction, training configuration, checkpoint selection, or other implementation details.

### Observed issue
The results I obtain are significantly lower than the paper-reported results.

### My setup
- Dataset: IEMOCAP
- Split: 81% train, 9% val, 10% test (default)
- Config used: hubert_base.py
- Training command:
  python train.py -cfg <config_path>/hubert_base.py

- Evaluation command:
  python eval.py -ckpt <checkpoint_path> -cm

### Results
- Paper-reported result: W-Acc: 79.25, UA-Acc: 78.92
- My reproduced result: W-Acc: 71.12, UA-Acc: 71.49

### Environment
- OS: Linux 5.14.0-427.91.1.el9_4.x86_64 (Linux-5.14.0-427.91.1.el9_4.x86_64-x86_64-with-glibc2.10)
- Python: 3.8.19
- PyTorch: 2.0.1
- CUDA (PyTorch build): 11.8
- CUDA available: True
- Transformers: 4.29.2
- NumPy: 1.24.4
- scikit-learn: 1.2.2
- torchaudio: 2.0.2
- torchvision: 0.15.2
- matplotlib: 3.7.1
- seaborn: 0.13.2

### Things I checked
- I used the public repository implementation
- I used the provided baseline configuration
- I evaluated the saved checkpoint on the test set
- I verified that training completed successfully
- I did not intentionally modify the baseline architecture

### Question
Are there any undocumented preprocessing steps, dataset filtering rules, environment/version requirements, or training details required to match the paper results?

I’d appreciate any clarification on:
1. the exact preprocessing expected for the reported results
2. whether the paper results were produced with the same public code
3. whether there are specific package/version constraints needed for reproduction
4. where I could get the pre-trained model if it exists, the releases are empty

Here is a log of the config, the training and eval runs, and the confusion matrix:

[cfg.log](https://github.com/user-attachments/files/25981121/cfg.log)
[train.log](https://github.com/user-attachments/files/25981136/train.log)
[eval.txt](https://github.com/user-attachments/files/25981106/eval.txt)

<img width="9600" height="6600" alt="Image" src="https://github.com/user-attachments/assets/ad3499c3-b6a5-48bc-bb6a-1bb6974c7b92" />

Thank you so much!

**Comment by uzecho** [comment]
I have the same problem

**Comment by tpnam0901** [comment]
Hi @CaydenSchalk @uzecho, thank you for your interest in our work. I will ask my colleague to check the code on the weekend. Could you please try running this repo to see if your environment can achieve similar performance for BERT and HuBERT, as shown in the figure below?
https://github.com/tpnam0901/4M-MER/blob/main/src/configs/4m-ser_bert_hubert.py 

<img width="918" height="298" alt="Image" src="https://github.com/user-attachments/assets/d63963fa-87d5-4d20-92ce-a4681359b111" />

**Comment by CaydenSchalk** [comment]
@tpnam0901  thank you for your response.

I set up the repository on our HPC cluster and trained the model on the IEMOCAP dataset following the README instructions.

My results were:

Validation set:
- BACC: 82.24
- ACC: 81.73
- MF1: 82.17
- WF1: 81.73

Test set:
- BACC: 76.07
- ACC: 75.81
- MF1: 76.37
- WF1: 75.91

The only environment-related change I made was downgrading mkl to 2024.0.0 to avoid the following import error:

`undefined symbol: iJIT_NotifyEvent`

I also changed the dataset setting in the config to IEMOCAP, since I assumed that was the dataset used for the ablation table.

Here are the eval logs, config, and training log:

[evaltestset.txt](https://github.com/user-attachments/files/26072432/evaltestset.txt)

[evalvalset.txt](https://github.com/user-attachments/files/26072433/evalvalset.txt)

[cfg.log](https://github.com/user-attachments/files/26072436/cfg.log)

[train.log](https://github.com/user-attachments/files/26072438/train.log)

As well as the slurm scripts I'm using to start the jobs:

[run_train.txt](https://github.com/user-attachments/files/26072502/run_train.txt)

[run_eval.txt](https://github.com/user-attachments/files/26072503/run_eval.txt) (this is for the val set, there was no -t parameter when I evaluated on the test set)

Thank you!

**Comment by tpnam0901** [comment]
Thank you. I think these is no problem with your environment. Could you try this modification from 4m-ser repo:
https://github.com/tpnam0901/4M-MER/blob/0051a66a352ff4673be1869b1055628d849c3644/src/models/networks.py#L126
In this line, please add min fusion head from this:
https://github.com/tpnam0901/MemoCMT/blob/d3248703e755319736b805fd6b0c2e8ab6089160/src/models/networks.py#L128
```
# Get classification output
if self.fusion_head_output_type == "cls":
    cls_token_final_fusion_norm = fusion_norm[:, 0, :]
elif self.fusion_head_output_type == "mean":
    cls_token_final_fusion_norm = fusion_norm.mean(dim=1)
elif self.fusion_head_output_type == "max":
    cls_token_final_fusion_norm = fusion_norm.max(dim=1)[0]
elif self.fusion_head_output_type == "min":
    cls_token_final_fusion_norm = fusion_norm.min(dim=1)[0]
else:
    raise ValueError("Invalid fusion head output type")
```

Then change the config to 'min' option:
https://github.com/tpnam0901/4M-MER/blob/0051a66a352ff4673be1869b1055628d849c3644/src/configs/base.py#L191
```
self.fusion_head_output_type: str = "min"  # [cls, mean, max]
```

**Comment by uzecho** [comment]
I think it is really dependent on the splits, since they are random, results are different. I used the same env and my results were:

Val set: BACC | ACC | MACRO_F1 | WEIGHTED_F1 
             76.94 & 76.10 & 76.66 & 76.10
Test set: BACC | ACC | MACRO_F1 | WEIGHTED_F1 
             78.52 & 77.44 & 78.15 & 77.52

As you can see compared to @CaydenSchalk 's result, I have lower val accuracy, but better test set accuracy. But still I couldn't even reach 80% accuracy.

**Comment by tpnam0901** [comment]
@uzecho, I think the difference might be due to this one.
https://github.com/tpnam0901/MemoCMT/blob/d3248703e755319736b805fd6b0c2e8ab6089160/scripts/preprocess.py#L75
The glob function returns different results in different environments.
The shuffle and split functions use the same seed, so I think it will be the same across environments.  

Btw, you can try with this preprocess:
https://github.com/tpnam0901/4M-MER/releases/download/v0.1.0/data_preprocessed.zip

**Comment by tpnam0901** [comment]
@CaydenSchalk @uzecho The preprocessing.py file differs from the one in the paper. I checked my PC, but I didn't have the backup for this repo. I will ask my colleague to upload the preprocessing.py file. From my autosuggestion in zsh history, the correct preprocess should be
```
python preprocess.py -ds IEMOCAP -dr data/dataset/IEMOCAP_raw/IEMOCAP_full_release --five_fold
```

**Comment by tpnam0901** [comment]
Hi @CaydenSchalk @uzecho, unfortunately, I already asked my colleague, and we do not have the backup for this repo. The code provided in this repo is our initial code when we submitted the paper. I deleted this repo last year, thinking I had already made all the commits. So far, as I remember, for the first stage, we perform a five-fold split using StratifiedKFold from sklearn, or just sort the data and then get the five-fold split. During revision, we also run additional experiments and modify the code, but I can't remember whether any changes were made to the model source code in this phase (I think no changes were made, only to the preprocessing part).  I am sorry for the out-of-date code. 

I think you can use this approach for running other experiments.
https://github.com/tpnam0901/MemoCMT/issues/2#issuecomment-4079085149

Thank you for your interest in our work.
